core/capabilities/compute/compute.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	id := generateID(binary)
+
+	m, ok := c.modules.get(id)
+	if !ok {
+		mod, err := c.initModule(id, binary, request.Metadata.WorkflowID, request.Metadata.ReferenceID)
+		if err != nil {
+			return capabilities.CapabilityResponse{}, err
+		}
+
+		m = mod
+	}
+
+	return c.executeWithModule(m.module, config, request)
+}
+
+func (c *Compute) initModule(id string, binary []byte, workflowID, referenceID string) (*module, error) {
+	initStart := time.Now()
+	mod, err := host.NewModule(&host.ModuleConfig{Logger: c.log}, binary)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate WASM module: %w", err)
+	}
+
+	mod.Start()
+
+	initDuration := time.Since(initStart)
+	computeWASMInit.WithLabelValues(workflowID, referenceID).Observe(float64(initDuration))
+
+	m := &module{module: mod}
+	c.modules.add(id, m)
+	return m, nil
+}
+
+func (c *Compute) popBytesValue(m *values.Map, key string) ([]byte, error) {
+	v, ok := m.Underlying[key]
+	if !ok {
+		return nil, fmt.Errorf("could not find %s in map", key)
+	}
+
+	vb, ok := v.(*values.Bytes)
+	if !ok {
+		return nil, fmt.Errorf("value is not bytes: %s", key)
+	}
+
+	delete(m.Underlying, key)
+	return vb.Underlying, nil
+}
+
+func (c *Compute) executeWithModule(module *host.Module, config []byte, req capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	executeStart := time.Now()
+	capReq := capabilitiespb.CapabilityRequestToProto(req)
+
+	wasmReq := &wasmpb.Request{
+		Id:     uuid.New().String(),
+		Config: config,
+		Message: &wasmpb.Request_ComputeRequest{
+			ComputeRequest: &wasmpb.ComputeRequest{
+				Request: capReq,
+			},
+		},
+	}
+	resp, err := module.Run(wasmReq)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	cresppb := resp.GetComputeResponse().GetResponse()
+	if cresppb == nil {
+		return capabilities.CapabilityResponse{}, errors.New("got nil compute response")
+	}
+
+	cresp, err := capabilitiespb.CapabilityResponseFromProto(cresppb)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, errors.New("could not convert response proto into response")
+	}
+
+	computeWASMExec.WithLabelValues(
+		req.Metadata.WorkflowID,
+		req.Metadata.ReferenceID,
+	).Observe(float64(time.Since(executeStart)))
+
+	return cresp, nil
+}
+
+func (c *Compute) Info(ctx context.Context) (capabilities.CapabilityInfo, error) {
+	return capabilities.NewCapabilityInfo(
+		CapabilityIDCompute,
+		capabilities.CapabilityTypeAction,
+		"WASM custom compute capability",
+	)
+}
+
+func (c *Compute) Start(ctx context.Context) error {
+	c.modules.start()
+	return c.registry.Add(ctx, c)
+}
+
+func (c *Compute) Close() error {
+	c.modules.close()
+	return nil
+}
+
+func NewAction(log logger.Logger, registry coretypes.CapabilitiesRegistry) *Compute {
+	compute := &Compute{
+		log:      logger.Named(log, "CustomCompute"),
+		registry: registry,
+		modules:  newModuleCache(clockwork.NewRealClock(), 1*time.Minute, 10*time.Minute, 3),
```

Comment:
just checking, is it alright to have cache config as a static thing?

Reply:
It is for now, there's a follow up ticket to source them from somewhere properly: https://smartcontract-it.atlassian.net/browse/CAPPL-95